core/src/main/kotlin/org/evomaster/core/problem/rest/service/AbstractRestFitness.kt

```diff
@@ -0,0 +1,465 @@
+                    .filter { name ->
+                        ! action.parameters.any { it is HeaderParam && it.name.equals(name, ignoreCase = true) }
+                    }
+                    .forEach {
+                        action.parameters.add(HeaderParam(it, OptionalGene(it, StringGene(it), false)))
+                    }
+
+            info.queryParameters
+                    .filter { name ->
+                        ! action.parameters.any { it is QueryParam && it.name.equals(name, ignoreCase = true) }
+                    }
+                    .forEach { name ->
+                        action.parameters.add(QueryParam(name, OptionalGene(name, StringGene(name), false)))
+                    }
+        }
+    }
+
+
+
+    /**
+     * Initializing Actions before evaluating its fitness if need
+     */
+    open fun doInitializingActions(ind: T) {
+    }
+
+    /**
+     * Create local targets for each HTTP status code in each
+     * API entry point
+     */
+    protected fun handleResponseTargets(
+            fv: FitnessValue,
+            actions: MutableList<RestAction>,
+            actionResults: MutableList<ActionResult>) {
+
+        (0 until actionResults.size)
+                .filter { actions[it] is RestCallAction }
+                .filter { actionResults[it] is RestCallResult }
+                .forEach {
+                    val status = (actionResults[it] as RestCallResult)
+                            .getStatusCode() ?: -1
+                    val name = actions[it].getName()
+
+                    //objective for HTTP specific status code
+                    val statusId = idMapper.handleLocalTarget("$status:$name")
+                    fv.updateTarget(statusId, 1.0, it)
+
+                    /*
+                        Objectives for results on endpoints.
+                        Problem: we might get a4xx/5xx, but then no gradient to keep sampling for
+                        that endpoint. If we get 2xx, and full coverage, then no gradient to try
+                        to keep sampling that endpoint to get a 5xx
+                     */
+                    val okId = idMapper.handleLocalTarget("HTTP_SUCCESS:$name")
+                    val faultId = idMapper.handleLocalTarget("HTTP_FAULT:$name")
+
+                    //OK -> 5xx being better than 4xx, as code executed
+                    //FAULT -> 4xx worse than 2xx (can't find bugs if input is invalid)
+                    if (status in 200..299) {
+                        fv.updateTarget(okId, 1.0, it)
+                        fv.updateTarget(faultId, 0.5, it)
+                    } else if (status in 400..499) {
+                        fv.updateTarget(okId, 0.1, it)
+                        fv.updateTarget(faultId, 0.1, it)
+                    } else if (status in 500..599) {
+                        fv.updateTarget(okId, 0.5, it)
+                        fv.updateTarget(faultId, 1.0, it)
+                    }
+                }
```

Comment:
I recently made some changes in the RestFitness class. Make sure those will end up here as well

Reply:
checked and merged